### DISCUSSION TITLE
Compiling ruby 3.1.3 failed! (MacOS Ventura 13.5.1 M1 Max)

### DISCUSSION THREAD

**Discussion by cameronhu** [rbenv/ruby-build#2245] [Build failures]
Here are the last log messages

<pre>
ossl_ts.c:829:5: error: incomplete definition of type 'struct TS_verify_ctx'
    TS_VERIFY_CTX_set_certs(ctx, x509inter);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./openssl_missing.h:215:46: note: expanded from macro 'TS_VERIFY_CTX_set_certs'
#  define TS_VERIFY_CTX_set_certs(ctx, crts) TS_VERIFY_CTS_set_certs(ctx, crts)
                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./openssl_missing.h:195:52: note: expanded from macro 'TS_VERIFY_CTS_set_certs'
#  define TS_VERIFY_CTS_set_certs(ctx, crts) ((ctx)->certs=(crts))
                                              ~~~~~^
/opt/homebrew/Cellar/openssl@3/3.1.2/include/openssl/ts.h:407:16: note: forward declaration of 'struct TS_verify_ctx'
typedef struct TS_verify_ctx TS_VERIFY_CTX;
               ^
1 warning and 1 error generated.
make[2]: *** [ossl_ts.o] Error 1
make[1]: *** [ext/openssl/all] Error 2
make: *** [build-ext] Error 2
!!! Compiling ruby 3.1.3 failed!
</pre>



I've already attempted uninstalling openssl@3 through Homebrew, and reinstalling using the code in the Suggested Build Environment section:

``` 
brew install openssl@3 readline libyaml gmp
export RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@3)"
```
I've also uninstalled Anaconda as was suggested in the troubleshooting section. Any help would be greatly appreciated!

**Comment by nroose** [comment]
Why is this closed?

**Comment by viperchaos** [comment]
You can try this below, I just solved the problem by the way below

The error you're encountering is due to an issue with the OpenSSL library version compatibility with the Ruby version you're trying to install. In this case, it seems like Ruby is having issues with the version of OpenSSL installed on your Mac (version 3.1.2).

Here are the steps you can follow to solve this issue:

### Step 1: Install Homebrew (if not already installed)

Homebrew is a package manager for macOS that can help you install and manage software packages easily. If you do not have it installed, you can [install it by following the instructions here](https://brew.sh).

### Step 2: Install a Compatible OpenSSL Version

You should try installing an older version of OpenSSL that is compatible with the Ruby version you're installing. You can do this using Homebrew:

```sh
brew install openssl@1.1
```

### Step 3: Set the OpenSSL Environment Variable

Next, you need to set the OpenSSL environment variable to point to the older version you just installed. This will help Ruby to find the correct version of OpenSSL during the installation. You can set the environment variable using the following commands:

```sh
export RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)"
```

### Step 4: Install Ruby

Now, try installing Ruby again. You can do this using a version manager like `rbenv` or `rvm`. Here, I'll demonstrate using `rbenv`:

1. Install `rbenv`:

   ```sh
   brew install rbenv
   ```

2. Add `rbenv` to bash so that it runs every time your session starts:

   ```sh
   echo 'if which rbenv > /dev/null; then eval "$(rbenv init -)"; fi' >> ~/.bash_profile
   source ~/.bash_profile
   ```

3. Install Ruby:

   ```sh
   rbenv install 3.1.3
   ```

### Step 5: Verify Ruby Installation

After installing Ruby, you can verify that it was installed successfully by running:

```sh
ruby -v
```

### Step 6: Install Jekyll

Once Ruby is installed, you can install Jekyll by running:

```sh
gem install jekyll
```

### Step 7: Verify Jekyll Installation

Verify that Jekyll was installed successfully by running:

```sh
jekyll -v
```

These steps should help you resolve the error and get Ruby and Jekyll installed on your system. Let me know if you face any issues or have further questions.

**Reply by kiumarsj** [reply]
I got this after installing openssl@1.1

configure: error: something wrong with LDFLAGS="-L/usr/local/opt/openssl@1.1/lib"
!!! Configuration of ruby 3.1.3 failed!

**Reply by DavidLozzi** [reply]
same here :(

**Reply by DavidLozzi** [reply]
Got it working with the following

```
export PATH="/opt/homebrew/opt/openssl@1.1/bin:$PATH"
export PKG_CONFIG_PATH="/opt/homebrew/opt/openssl@1.1/lib/pkgconfig"
export LDFLAGS="-L/opt/homebrew/opt/openssl@1.1/lib"
export CPPFLAGS="-I/opt/homebrew/opt/openssl@1.1/include"
```

and then continue the steps

```
ruby-install ruby 3.1.3
```

**Reply by kiumarsj** [reply]
thanks it worked 😃

**Reply by ryankuykendall** [reply]
Note that these should be prefixed with $HOME if you are using the default Homebrew install:

export PATH="$HOME/homebrew/opt/openssl@1.1/bin:$PATH"
export PKG_CONFIG_PATH="$HOME/homebrew/opt/openssl@1.1/lib/pkgconfig"
export LDFLAGS="-L$HOME/homebrew/opt/openssl@1.1/lib"
export CPPFLAGS="-I$HOME/homebrew/opt/openssl@1.1/include"

**Reply by mislav** [reply]
When linking to Homebrew's OpenSSL, please avoid pasting commands that worked for you but that might not work for other people.

The Homebrew installation prefix is different across different macOS architectures (Intel vs. Apple Silicon) and can be different based on the user's settings. Instead of hardcoding a path like `/opt/homebrew/...`, you can invoke the command `brew --prefix openssl@1.1` and use the result of that:

```sh
brew install openssl@1.1
export RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)"
rbenv install 3.1.3
```
Note that putting extra entries in PATH, LDFLAGS, or CPPFLAGS is absolutely *not* necessary when compiling Ruby. For more information, please [refer to our wiki](https://github.com/rbenv/ruby-build/wiki).

**Reply by MichalAFerber** [reply]
This worked for me as well.

**Reply by DaraDadachanji** [reply]
this worked for me except I also had to run

``` 
rbenv global 3.1.3
source ~/.bash_profile
```

in between steps 4 and 5

**Reply by rrshaban** [reply]
Is there any danger using (and hard-coding) openssl@1.1 when the current version is openssl@3 3.3.0 ?

**Comment by nroose** [comment]
I mean, seems like at the very least, most environments need PATH set. And
your instructions are very different from the ones on the page you linked
to. And both sets of instructions seem to not really work for me. I have to
intentionally uninstall OpenSSL3 and ignore those dependencies and then
only OpenSSL 1.1 is installed and then I can build Ruby 3.2.2.

**Reply by mislav** [reply]
> And both sets of instructions seem to not really work for me. I have to intentionally uninstall OpenSSL3 and ignore those dependencies and then only OpenSSL 1.1 is installed and then I can build Ruby 3.2.2.

That is odd, since Ruby 3.2.2 should definitely be compatible with OpenSSL 3. You could check or share the full build log to find out more about what happened when trying to link to Homebrew's `openssl@3`. I don't think the solution should be to uninstall it— that was only necessary for Ruby 2.6, but not for any Ruby version after that.

Here is my local Ruby 3.2.2:
```
ruby -r openssl -e 'p RUBY_VERSION, OpenSSL::OPENSSL_VERSION'
"3.2.2"
"OpenSSL 3.1.4 24 Oct 2023"
```
